docs/specs/code/SIGNING.md

```diff
@@ -0,0 +1,183 @@
+
+lastly, an array of witnesses is represented by a fixed-length CBOR array (call this function `FIXED_ARRAY`)
+
+```rust
+serializer.write_array(Len::Len(data.len() as u64))?;
+for element in data {
+    Serialize::serialize(element, serializer)?;
+}
+```
+
+You can see all existing tags in Rust [here](https://github.com/input-output-hk/rust-cardano/blob/2e924a4457203fa6b19f409cb897d90cd6b699f6/cardano/src/tags.rs) and in Haskell [here](https://github.com/input-output-hk/cardano-sl/blob/ccca8b346413f12e9fad8cab4c855e69669a2151/crypto/Pos/Crypto/Signing/Tag.hs#L15)
+
+### Proposed format
+
+##### Definitions
+
+**Define**: New CBOR tag `MessageTag = 0x0b` to be used for our new message signing standard. This tag is different than all other tags for meaningful signed data inside Cardano.
+
+**Define**: byte array `UserMessage` which is the message the user wishes to sign. It is RECOMMENDED to use unicode with utf8 encoding if the message is a string and RECOMMENDED to use network-byte-order for numbers.
+
+**Define**: function `BASE32` which converts bytes to base32 in the following way
+
+1) Consider byte array as an array of 5-bit values (most significant bit first)
+1) 0-pad last item if necessary
+1) Translate the array elements using the [bech32 table](https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki#bech32)
+
+**Define**: uint8 `KeyType` which exists so that programming languages can assign the correct type to these keys if needed. Although there is no standard specifying a number for each key type, in practice the following are used
+
+| Type            | ID |
+|-----------------|----|
+| XPub            | 0  |
+| ValidatorScript | 1  |
+| RedeemPublicKey | 2  |
+
+**Define**: uint32 `ProtocolMagic` which specifies the chain id. Notably, the `ProtocolMagic` for the Cardano Mainnet is `764824073`
+
+**Define**: function `SIGN` which cryptographically signs data. This function is purposely left vague as the implementation depends on what kind of key is used.
+
+**Define**: uint8 `FormatVersion` which specifies the version number for the proposed format.
+
+**Define**: byte* `Recipient` which is a identifier for who you are signing this message for. This is to avoid replay attacks as dapps for example can specify their contract address as the recipient. It is still not fool-proof as somebody can ask you to sign a message with the recipient set the same as the recipient field of the contract/person they intend to pass your message along to.
+
+##### FormatVersion 1
+
+```
+HRP             ::= 'cm'
+UserMessage     ::= byte*
+MessageTag      ::= 0x0b
+ProtocolMagic   ::= uint32
+Recipient       ::= byte*
+Signature       ::= SIGN(MessageTag || ProtocolMagic || UserMessage)
+                  | SIGN(MessageTag || ProtocolMagic || Recipient || UserMessage)
```

Comment:
Yes I intended `||` to mean byte concatenation.

They don't have to know. The presence of a recipient is just mean to avoid collisions where two apps ask you to sign the same string. Maybe you missed that all this is wrapped by `SIGN` so who you signed the message for cannot be extracted by a 3rd party. There is some privacy benefit but mostly I didn't want to attach the recipient as plain text because it would increase the length of the bech32 string.

A wallet UI would have two input fields: `recipient` and `message` and the code would put these together to form the bytestring you need to sign. 
The recipient knows whether or not the recipient field should have been included so they know what bytestring was signed.

Reply:
Wouldn't be possible to check for the recipient field by decoding the raw data `BASE32([ FormatVersion , Elements, Recipient ])`? If not, isn't repetitive including the recipient in both the signature and the raw data?